### PR TITLE
TGR-102: Updating ruby to 3.3, docs, and minor aws gems implementation

### PR DESCRIPTION
**Pull request by EdwinGuzman**
Resolves [TGR-102](https://newyorkpubliclibrary.atlassian.net/browse/TGR-102)

While I got the repo to work locally following the instructions in the readme using ruby 3.3, I'm not entirely sure how to go about testing this in QA or how other services are related to this.

- I'm not familiar with the ruby and gem ecosystem so please let me know if I need to make any updates in the gem files.
- Tests run and pass.
- Localstack had minor updates since this was last updated. I had to create a new encrypted `SQS_QUEUE_URL` variable to use port "4566" instead of "4567".
- Travis is replaced by Github Actions for running tests and the deployment process.

[TGR-102]: https://newyorkpubliclibrary.atlassian.net/browse/TGR-102?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,7 +1,8 @@
 source 'https://rubygems.org'
 
-gem 'aws-sdk-sqs'
-gem 'aws-sdk-kms'
+gem 'aws-sdk-sqs', '~> 1.81'
+gem 'aws-sdk-core', '~> 3.203'
+gem 'aws-sdk-kms', '~> 1.89'
 gem 'nypl_log_formatter', '~> 0.1.2'
```
Comment (croyfish) [inline Gemfile:6]: These upgrades seem reasonable. What I have been doing (for example, with [mms](https://github.com/NYPL/mms/blob/qa/Gemfile)) is leaving required versions off entirely, removing Gemfile.lock, and then regenerating it with `bundle install` to get the latest versions. 

By not adding version constraints, we should be able to keep dependencies up to date without having to think about it too much, and if we find a case where a version needs to be locked, then we can do it at that point.

I was wondering a bit about nypl_log_formatter. It looks like we could actually upgrade it to [version 0.1.3](https://rubygems.org/gems/nypl_log_formatter/versions), but that dependency itself hasn't been updated in many years (edit: I see in the Gemfile.lock that we are pulling in 1.3). nypl_platform_api_client is likewise > 4 years old, but we're already on the latest version of that.

**Review by croyfish — APPROVED** [review]
Went through the local steps as well, and it seems to be working with the proposed changes:

```
2024-09-11T15:47:23.021  INFO --- [et.reactor-0] localstack.request.aws     : AWS sqs.ReceiveMessage => 200
2024-09-11T15:49:01.793  INFO --- [et.reactor-0] localstack.request.aws     : AWS sqs.ReceiveMessage => 200
```

Similarly, `bundle exec rspec` shows no failures. You're right that other expected behavior is not well documented here, and not having everything running in docker containers is a little rough, but on the whole, I think we should be fine with these upgrades.

**Comment by EdwinGuzman** [comment]
@croyfish thanks for the review. I just removed the "WIP" and replaced travis with github actions and terraform. It'll have to be merged into `qa` in order to test that the deployment works but this follows the same pattern as other LSP lambdas.

File: Gemfile (hunk 1 above)
Comment (EdwinGuzman) [reply Gemfile:6]: I removed the versions from the aws-related gems but left it for the nypl-related gems.

File: provisioning/base/resources.tf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    error_message = "The environment must be 'qa' or 'production'."
+  }
+}
+
+# Package the app as a zip:
+data "archive_file" "lambda_zip" {
+  type        = "zip"
+  output_path = "${path.module}/dist.zip"
+  source_dir  = "../../"
+  excludes    = [".git", ".terraform", "provisioning"]
```
Comment (charmingduchess) [inline provisioning/base/resources.tf:21]: probably can include spec files in this array?

**Review by charmingduchess — COMMENTED** [review]
Looks good to me! I also do not have much insight about the ruby upgrade details.

**Review by charmingduchess — APPROVED** [review]

File: .github/workflows/test-and-deploy.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      - run:
+          bundle exec rspec
+
+  deploy-lambda-qa:
+    runs-on: ubuntu-latest
+    permissions: 
+      id-token: write
+      contents: read
+    needs: test
+    if: github.ref == 'refs/head/qa'
```
Comment (danamansana) [inline .github/workflows/test-and-deploy.yml:21]: I've been using `tgr-qa` as a temporary measure but this is fine if you are comfortable using the regular `qa` branch

File: provisioning/base/resources.tf [hunk 4]
```diff
@@ -0,0 +1,50 @@
+provider "aws" {
+  region = "us-east-1"
+}
+
+variable "environment" {
```
Comment (danamansana) [inline provisioning/base/resources.tf:5]: How is this app handling environment variables? Other lambdas that use terraform use something like this: 

`# Load ENV vars from ./config/{environment}.env
  environment {
    variables = { for tuple in regexall("(.*?)=(.*)", file("../../config/${var.environment}.env")) : tuple[0] => tuple[1] }
  }`

**Review by danamansana — APPROVED** [review]
A couple of minor comments/questions but looks good

File: provisioning/base/resources.tf (hunk 4 above)
Comment (nonword) [reply provisioning/base/resources.tf:5]: Looks like previously app config was only stored in `.travis.yml`, which is removed in this PR. The env vars should prob be moved to `./config/*.env` files and loaded in tf as @danamansana suggests above.

File: README.md [hunk 5]
```diff
@@ -70,14 +70,15 @@ aws_secret_access_key =
...
+```sh
+aws sqs create-queue --region us-east-1 --queue-name proxy-request-service --endpoint http://localhost:4566 --profile local
 ```
 
 When populating an SQS queue, the `aws sqs` cli tool may be useful for inspecting the messages written. For example, when populating a localstack SQS, run the following to pop the last 10 messages:
 
-```
-aws sqs receive-message --region us-east-1 --queue-url http://localhost:4576/queue/proxy-request-service --endpoint http://localhost:4576 --profile local --attribute-names All --message-attribute-names All --max-number-of-messages 10
+```sh
+aws sqs receive-message --region us-east-1 --queue-url http://localhost:4566/000000000000/proxy-request-service --endpoint http://localhost:4566 --profile local --attribute-names All --message-attribute-names All --max-number-of-messages 10
```
Comment (nonword) [inline README.md:81]: Is that `000000000000` intentional?

Comment (EdwinGuzman) [reply README.md:81]: Yes, that's the new url structure from localstack

File: .github/workflows/test-and-deploy.yml (hunk 3 above)
Comment (EdwinGuzman) [reply .github/workflows/test-and-deploy.yml:21]: I think this should be ok to merge directly and then test or do you prefer testing the PR and then merging it? We'd have to update the lambda to use the ruby 3.3 runtime, but if it fails, I don't think we can revert to a lower ruby runtime.

**Comment by EdwinGuzman** [comment]
Updated the env vars in terraform

**Review by nonword — APPROVED** [review]
Looks right to me. I can assist testing in QA if you like.

**Review by charmingduchess — APPROVED** [review]

File: lib/kms_client.rb [hunk 6]
```diff
@@ -3,7 +3,11 @@
 
 class KmsClient
   def initialize
-    @kms = Aws::KMS::Client.new(region: 'us-east-1')
+    @kms = Aws::KMS::Client.new(
+      region: 'us-east-1',
+      access_key_id:  ENV['AWS_ACCESS_KEY_ID'],
+      secret_access_key: ENV['AWS_SECRET_ACCESS_KEY']
+    )
```
Comment (croyfish) [inline lib/kms_client.rb:10]: Was this change a deliberate choice, or did you find that it was required?

Comment (EdwinGuzman) [reply lib/kms_client.rb:10]: Typically, the aws creds from the `~/.aws/credentials` file are used implicitly when the app runs. I found that in this docker setup, that wasn't true and I had to pass them in explicitly.

**Pull request merged**